### ISSUE TITLE
It should be possible to configure Open Browser keyword to not always open a new browser.

### ISSUE THREAD

**Issue by aaltat** [robotframework/SeleniumLibrary#1319]
Many times, when opening browser, there is need to open only one browser. But because depending how test are executed in Robot Framework, it is hard to find a single correct place for the `Open Browser` keyword. Most of the time, it is safer to run the `Open Browser` keyword from multiple levels in the suite structure, which will lead to many open browsers and loss in test execution time. There are also numerous other workarounds available. 

Instead of the workarounds, it would be good to implement support for the `Open Browser` keyword  to open browser only once. If the the `Open Browser` keyword is configured to open browser only one time, it would act in same way as the `Go To` keyword. 

The `alias` argument could act as a marker, should the existing browser be used or new one opened. 

Currently if test does this: 
```
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome | alias |
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome | alias |
```
It will open two browsers with same alias. 

But after the change, this:
```
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome | alias |
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome | alias |
```
will open one browser. With the first `Open Browser` keyword run, it will open the browser normally and navigate to the `https://github.com/robotframework/SeleniumLibrary`. In the second `Open Browser` keyword run, it navigate (again) to the `https://github.com/robotframework/SeleniumLibrary`. 

Also this:
```
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome | alias |
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Firefox | alias |
```
will open only one browser.

If different alias is used: 
```
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome | alias1 |
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome | alias2 |
```
it will open two browsers. 

Also if alias is not defined
```
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome |
| Open Browser | https://github.com/robotframework/SeleniumLibrary | Chrome |
```
it will open two browsers. 

There is small change that someone is misusing the `alias` argument and therefore this may cause problems for users. But because the problem should be easy to fix and the probability for that is quite low, implementing this is OK. 

This idea is also presented in #701

**Comment by rasjani** [comment]
Code change for this looks very minimal - however there's one gotcha.

in  keywords/browsermanagement, when calling open browser, a new driver is created and then registered into robot.utils.ConnectionCache. Alias can be used to fetch the already existing driver from it via get_connection() which supports both index and alias.  And here is the gotcha: current implementation returns the index of the driver that's being registered.  Now, if that feature must remain, there needs to be a way to also fetch the index but atm on the ConnectionCache, that feature is private (`_resolve_alias_or_index`). 

Ways forward ? 

 * Call  _resolve_alias_or_index anyway ? 
 * Use internal lookup table that maps alias names to their index
 * Possible changes to robot.utils.ConnectionCache  for getting the index of alias ?

**Comment by aaltat** [comment]
I haven't think about where the code resolving could go. But because we support also older Robot Framework versions, changing the RF code does not resolve the problem for short term. But by looking at SeleniumLibrary code, I think that it would be best to implement the lookup in the `WebDriverCache`. I would avoid relying too much the RF side internal methods, but then again if it makes SeleniumLibrary code easier to read, then reinventing wheel is not required. Usually only there is only few browsers open, then implementing internal lookup table doesn't not consume that much extra resources.

As a bottom line, it can be implemented in both ways and I don't have strong opinion which way it should be implemented. If you want, you can give it a try and if it's readable I am happy with it

**Comment by pekkaklarck** [comment]
I'd say there are several good solutions:

1. Call the private method in ConnectionCache. But also submit an issue/PR exposing that method and then change SL to use the new method once it's available. Some work but possibly benefits also others.

2. Have a separate index to alias map in SL. Easy.

3. Don't return the index in the first place. I don't see much use for them because aliases are a much better way to switch between browsers. One possibility would be returning the actual WedDriver instance instead. If that's done, `Switch Browser` could be enhanced to allow using that when switching making the change mostly backwards compatible. Should consider making it an error to use `Open Browser` twice without an alias, though.

**Comment by aaltat** [comment]
Well I would go with option 1or 2. If I would start doing the chance, I would first try with option 1 and see how does it look like.

**Comment by rasjani** [comment]
One alternative here:
https://github.com/rasjani/SeleniumLibrary/commit/4c306f57417f92e4534b17608fed5754d5d87bab

So, for this, _resolve_alias_or_index in rf side should also get changed to public. If this approach looks good, i'll proceed with that and update the documentation and write tests for this. Basic tests for this is already done but not committed but they dont really verify anything except what i can see in manual inspection.

Edit/Addition:  If same alias is used for 2 or more different browsers, how would that be addressed ? Issue a warning to the user or just bork out  ?

**Comment by aaltat** [comment]
By a quick look (I did spend 5 mins on reading it), I like it. For my side it is go for further implementation. 

Yes, Pekka was saying that `_resolve_alias_or_index ` could be changed as public method. Please note that for backwards compatibility reasons, so private method should also exist. But perhaps it should emit a warning. 

Same index can happen, no idea how common it might be. No idea how it actually works now, which browser is selects or is an exception raised? I will assume that it will select one of the browsers and in that case, displaying a warning is a good idea (warning when the browser is created at least, can there be warning in other scenarios? Well, one warning should be enough.) It would be too much of an hassle for users if we start failing without an deprecation period.

**Comment by rasjani** [comment]
Same alias issue i was referring to was something along the lines:

```
Open Browser    ${URL}    chrome    alias=mybrowser
Open Browser    ${URL}    firefox      alias=mybrowser
```

Then, current work in progress code would return chromedriver/chrome for second `Open Browser` keyword. That might be something that user would not expect but imho, its their own issue and minimal information about this scenario would be to issue a warning to the user that they are using same alias for multiple browsers.

Other comments, good that you pointed out the deprecation strategy .. Thanks!

**Comment by aaltat** [comment]
Hmmm, as long as we don't charge the how the browser is selected, when same alias is used, users will not have to make modifications to their tests because of this PR. And then they will be more happier users.

**Comment by rasjani** [comment]
In the last build all except 2.7 python build were passing.  Local build on pr branch had 3 failures on chrome. Same on master.  107 with Firefox.  So sorting out what is actually relevant and what’s broken feels a too much of a chore for now.  PR is here, when master gets stabile, I’ll take a new look

**Comment by aaltat** [comment]
I was thinking about different types of alias. Example if I have 
```
| Open Browser | ${URL} | chrome | alias=1 |
```
Then should this

```
| Open Browser | ${URL} | chrome | alias=${1} |
```
open a new browser or return existing one. I am leaning towards that it should open new browser, because numbers and strings are not the same thing.

**Comment by pekkaklarck** [comment]
I seriously doubt anyone has used non-string aliases and thus `alias=1` and `alias=${1}` isn't very relevant. On the other hand

```
    Switch Connection    1
```

can be ambiguous as that `1` could refer to an index (which actually is an integer) or an alias. If I remember correctly, `ConnectionCache` first considers the value an alias, so the above works if you've used `alias=1` earlier regardless was that the first opened browser or not. When it considers the value as an index, it first converts it to an `int`.

**Comment by pekkaklarck** [comment]
Somewhat related to above, and [as I commented already earlier](https://github.com/robotframework/SeleniumLibrary/issues/1319#issuecomment-472052551), I don't think it's important that `Open Connection` returns an index. For backwards compatibility reasons it should return *something* that can be used like

```
    ${something} =    Open Connection    # ...
    # ...
    Switch Connection    ${something}
```

but that something doesn't need to be an integer. Returning the underlying `WebDriver` instance would work as well and would also make it easier to do custom actions with it:

```
    ${driver} =    Open Connection    # ...
    Call Method    ${driver}    method    arg
```

Because `ConnectionCache` anyway supports indices, it ought to be easy to still support using indices that are passed as literal values like

```
    Open Connection    # ...
    # ...
    Switch Connection    1
```

Explicitly specifying an alias or using the value returned by `Open Connection` is much cleaner, though, and if the support for indices causes any extra work it could be deprecated.

**Comment by aaltat** [comment]
Opened because incomplete None handling.
